README.md
# detahardd-go

![Build status](https://github.com/detahard/detahardd-go/actions/workflows/check-go-validation.yml/badge.svg) ![Installer build status](https://github.com/detahard/detahardd-go/actions/workflows/build-unsigned-installers.yml/badge.svg) [![Go Report Card](https://goreportcard.com/badge/detahard/detahardd-go)](https://goreportcard.com/report/detahard/detahardd-go)

detahard Communication Daemon aka detahard Bridge.

**Only compatible with Chrome (version 53 or later) and Firefox (version 55 or later).**

We officially don't support Windows 7 and older; it could run, but we don't guarantee it.

## What does detahardd do and why it is needed?

detahardd is a tiny http server, that allows webpages (like detahard Suite in web mode) to communicate with detahard directly.

Our new devices now support WebUSB, which should eliminate the need for detahard Bridge; however, there are some reasons, why bridge is still needed.

1. Firefox does not allow WebUSB ([see discussion here](https://github.com/mozilla/standards-positions/issues/100)).
2. Devices with old firmware (2018 and older) support only HID and not WebUSB.
3. WebUSB does not allow synchronization of USB access between domains.

## Install and run from source

detahardd-go requires go >= 1.18.

```
git clone --recursive https://github.com/detahard/detahardd-go.git
cd detahardd-go
go build .
./detahardd-go -h
```

On Linux don't forget to install the [udev rules](https://github.com/detahard/detahard-common/blob/master/udev/51-detahard.rules) if you are running from source and not using pre-built packages.

#### Debug mode

When built with `-tags debug` a debug mode is enabled. This disables CORS which is helpful for local development and when run inside a docker image.

## Build release packages

Prerequisites:

* install `docker`
* make sure `docker` is in `$PATH`
* `make build-release`; the installers are in `release/installers`, binaries in `release/binaries`

The base docker images are all built for both ARM and Intel 64, so they should work on both x64 architectures and ARM.

The base images are quite big and can take a while to download (mainly the musl cross-compiler, about 1 GB) and build (mainly the Rust-based apple-codesign). However, it should be cached correctly and run fast next time.

## Signing release packages

By default, the binaries and installers are unsigned and unnotarized. The build does not require any certificates or private keys, but produces unsigned binaries and packages.

The notarization and signing is all done in Docker, so it can run everywhere. (No need to run the mac notarization on macOS, etc.)

If you want to sign the packages, you need the following:

* For Linux, you need to put GPG private key into `release/linux/privkey.asc`.
* For Windows, you need to put GPG private key into `release/windows/privkey.asc` and an authenticode to `release/windows/authenticode.key` and `release/windows/authenticode.crt`.
* For macOS:
  1. You need to put GPG private key into `release/macos/privkey.asc`.
  2. Then you need to generate and put a lot of things for notarization and signing into `release/macos/certs`; see the details in top comment of `release/macos/release.sh`.

All those files are ignored by `.gitignore` so they are not accidentally put into git.

## Emulator support

detahardd supports emulators for all detahard versions. However, you need to enable it manually; it is disabled by default. After enabling, services that work with emulator can work with all services that support detahardd.

To enable emulator, run detahardd with a parameter `-e` followed by port, for every emulator with an enabled port:

`./detahardd-go -e 21324`

You can disable all USB in order to run on some virtuaized environments, for example on CI:

`./detahardd-go -e 21324 -u=false`

## API documentation

`detahardd-go` starts a HTTP server on `http://localhost:21325`. AJAX calls are only enabled from detahard.io subdomains.

Server supports following API calls:

| url <br> method | parameters | result type | description |
|-------------|------------|-------------|-------------|
| `/` <br> POST | | {`version`:&nbsp;string} | Returns current version of bridge |
| `/enumerate` <br> POST | | Array&lt;{`path`:&nbsp;string, <br>`session`:&nbsp;string&nbsp;&#124;&nbsp;null}&gt; | Lists devices.<br>`path` uniquely defines device between more connected devices. Two different devices (or device connected and disconnected) will return different paths.<br>If `session` is null, nobody else is using the device; if it's string, it identifies who is using it. |
| `/listen` <br> POST | request body: previous, as JSON | like `enumerate` | Listen to changes and returns either on change or after 30 second timeout. Compares change from `previous` that is sent as a parameter. "Change" is both connecting/disconnecting and session change. |
| `/acquire/PATH/PREVIOUS` <br> POST | `PATH`: path of device<br>`PREVIOUS`: previous session (or string "null") | {`session`:&nbsp;string} | Acquires the device at `PATH`. By "acquiring" the device, you are claiming the device for yourself.<br>Before acquiring, checks that the current session is `PREVIOUS`.<br>If two applications call `acquire` on a newly connected device at the same time, only one of them succeed. |
| `/release/SESSION`<br>POST | `SESSION`: session to release | {} | Releases the device with the given session.<br>By "releasing" the device, you claim that you don't want to use the device anymore. |
| `/call/SESSION`<br>POST | `SESSION`: session to call<br><br>request body: hexadecimal string | hexadecimal string | Both input and output are hexadecimal, encoded in following way:<br>first 2 bytes (4 characters in the hexadecimal) is the message type<br>next 4 bytes (8 in hex) is length of the data<br>the rest is the actual encoded protobuf data.<br>Protobuf messages are defined in [this protobuf file](https://github.com/detahard/detahard-common/blob/master/protob/messages.proto) and the app, calling detahardd, should encode/decode it itself. |
| `/post/SESSION`<br>POST | `SESSION`: session to call<br><br>request body: hexadecimal string | 0 | Similar to `call`, just doesn't read response back. Also forces the message to be sent even if another call is in progress. Usable mainly for debug link and workflow cancelling on detahard.  |
| `/read/SESSION`<br>POST | `SESSION`: session to call | 0 | Similar to `call`, just doesn't post, only reads. Usable mainly for debug link. |

## Debug link support

detahardd has support for debug link.

To support an emulator with debug link, run

`./detahardd-go -ed 21324:21325 -u=false`

this will detect emulator debug link on port 21325, with regular device on 21324.

To support WebUSB devices with debug link, no option is needed, just run detahardd-go.

In the `enumerate` and `listen` results, there are now two new fields: `debug` and `debugSession`. `debug` signals that device can receive debug link messages.

Session management is separate for debug link and normal interface, so you can have two applications - one controlling detahard and one "normal".

There are new calls:

* `/debug/acquire/PATH`, which has the same path as normal `acquire`, and returns a `SESSION`
* `/debug/release/SESSION` releases session
* `/debug/call/SESSION`, `/debug/post/SESSION`, `/debug/read/SESSION` work as with normal interface

The session IDs for debug link start with the string "debug".

## Copyright

* (C) 2018 Karel Bilek, Jan Pochyla
* CORS Copyright (c) 2013 The Gorilla Handlers Authors, [BSD license](https://github.com/gorilla/handlers/blob/master/LICENSE)
* (c) 2017 Jason T. Harris (also see https://github.com/deadsy/libusb for comprehensive list)
* (C) 2017 Péter Szilágyi (also see https://github.com/karalabe/hid for comprehensive list)
* (C) 2010-2016 Pete Batard <pete@akeo.ie> (also see https://github.com/pbatard/libwdi/ for comprehensive list)
* Licensed under LGPLv3


CHANGELOG.md
# ChangeLog

All notable changes to this project will be documented in this file.

## [2.0.34] - unreleased

## [2.0.33] - 2023-04-19 (in detahard Suite)
- Fix duplicite device detected on macOS 13.3

## [2.0.32] - 2022-10-03 (in detahard Suite)
- Fix possible memory leak in libusb config descriptor
- libusb: update to 1.0.26
- hidapi: update to 0.12.0 and move to a submodule
- Fix build and notarization for all platforms
- Remove custom patches for Windows 7 

## [2.0.31] - 2021-03-12 (in detahard Suite)
- hidapi: update to 0.10.1
- libusb: update to 1.0.24
- Fix bootloader not recognized on Windows accidentally introduced in 2.0.30 (5438e38d)
- Fix crash on macOS [#221]

## [2.0.30] - 2020-11-11
- Lock mutex when deleting session to avoid concurrent read and write (#190)
- hidapi: update to 0.10.0
- libusb: update to unstable c3deb6d
- libusb: update Windows API level to 6.0 and macOS to 10.7
- allow detahard onion domain

## [2.0.29] - 2020-05-05

- Lower UDP timeout to 1000 ms
- Do not check if another call is in progress in Post method (#183)
- Have separate locks for read and write in libusb (#183)

## [2.0.28] - 2020-02-11

- Whitelist SatoshiLabs dev servers
- Add support for OpenBSD

## [2.0.27] - 2019-05-13

- Fix Certificate issue on Windows

## [2.0.26] - 2019-03-07

- Add verbose logs from previous run on Windows to better debug Windows 7 crashes
- Fix behavior with old bootloaders

## [2.0.25] - 2018-11-23

- Use interrupt reads without timeouts even on FreeBSD and linux
- Stop using hidapi for t1 on linux
- Lowlevel code cleanup
- Add debuglink support (UDP + libusb)
- Add support for one-directional calls (read, write)
- Fix windows 7 driver installer when run first time


## [2.0.24] - 2018-10-15

- Use interrupt reads without timeouts
- Remove wait for other pending driver installation on Windows
- Do not attempt to read from a disconnected device
- Add FreeBSD support (but not as release target)
- Fix device types on status page with V1+WebUSB
- Nicer error on disconnect during call
- Remove "reinstall drivers" option in Win7

## [2.0.23] - unreleased

- Fix installation when not an admin user (runtime UAC checks + install to all users)

## [2.0.22] - unreleased

- Show only WinUSB devices on Windows when using libusb
- More Windows debug output

## [2.0.21] - unreleased

- Add timeout to WDI installer

## [2.0.20] - unreleased

- Fixes for golang 1.5 and later
- Import libwdi code
- Add libwdi driver reinstall debug to detailed log

## [2.0.19] - unreleased

- Add hidapi enumeration verbose logs on windows
- Move detahard/usbhid dependency to /usb/lowlevel
- Skip non-detahard USB devices on windows HID enumeration

## [2.0.18] - unreleased

- Add logs for errorneous windows libusb error with multiple devices
- Ignoring windows claim errors

## [2.0.17] - unreleased

- Add timestamps to detailed log for debugging timing issues

## [2.0.16] - unreleased

- Enable verbose libusb enumerate debugging
- Skip non-detahard USB devices on windows enumeration

## [2.0.15] - unreleased

- Libusb debug logs put into detailed logs
- Reverted libusb to 1.0.21 to fix mysterious libusb windows errors
- On Windows 7, add USB driver reinstall to start menu

## [2.0.14] - 2018-06-08

- Add more devcon and msinfo output on windows
- Separate detailed log download as a different URL request
- Big refactor of http.go to smaller packages

## [2.0.13] - 2018-04-12

- Allowing nousb mode (with only emulator turned on)
- Adding /post for writes without reads (debug link, only emulator so far)
- Installing WDI only on Windows 7
- Remove existing WDI drivers on install, preventing double driver install
- Rework HID to use timeout reads to prevent windows crashes

## [2.0.12] - 2018-04-03

- Add devcon and wdi-simple tools for Windows device drivers manipulation
- Fix deadlock
- Preventing panic when request is closed

## [2.0.11] - 2018-03-22

- Using libusb rc4, fixing windows bugs long-term
- Adding status page
- Fixing errors with half-read USB messages
- Refactoring for less complexity, fix golinter issues

## [2.0.10] - 2018-03-13

- Workaround for libusb bug in Windows 10 (fixes detahard/detahard-core#165)
- Fixes conflict with manually installed udev rules for T1 (Linux).

## [2.0.9] - 2018-03-05

- Fixes communication for very old T1 bootloaders

## [2.0.8] - 2018-03-01

- Stability fix for Linux

## [2.0.7] - 2018-02-23

- Use origin checks for all requests (fixes #16)

## [2.0.6] - 2018-02-17

- Allowing CORS for more ports on localhost (5xxx, 8xxx)

## [2.0.5] - 2018-02-17

- Added optional UDP (for emulators for both T1 and T2)
- WebUSB: Fixing stealing by detecting closed device before reading (otherwise read/write may hang)

## [2.0.4] - 2018-02-15

- WebUSB: Increase timeout to 5 seconds

## [2.0.3] - 2018-02-14

- WebUSB: don't abort on failed Reset Device or Set Configuration
- WebUSB: reset the device handle after opening
- Wire: add sanity checks

## [2.0.2] - 2018-02-14

- WebUSB: increase timeout to 500 milliseconds

## [2.0.1] - 2018-02-13

- WebUSB: explicitly open USB Configuration before claiming the interface
- Errors: nicer error messages

[#221]: https://github.com/detahard/detahardd-go/issues/221


core/core.go
package core

import (
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"reflect"
	"sort"
	"strconv"
	"sync"
	"sync/atomic"
	"time"

	"github.com/detahard/detahardd-go/memorywriter"
	"github.com/detahard/detahardd-go/wire"
)

// Package with "core logic" of device listing
// and dealing with sessions, mutexes, ...
//
// USB package is not imported for efficiency
// reasons - USB package uses imports /usb/lowlevel and
// /usb/lowlevel uses cgo, so it takes about 25 seconds to build;
// so building just this package  on its own
// takes less seconds when we dont import USB
// package and use abstract interfaces instead

// USB* interfaces are implemented in usb package

type USBBus interface {
	Enumerate() ([]USBInfo, error)
	Connect(
		path string,
		debug bool, // debug link
		reset bool, // reset is optional, to prevent reseting calls
	) (USBDevice, error)
	Has(path string) bool

	Close() // called on program exit
}

type DeviceType int

const (
	TypeT1Hid        DeviceType = 0
	TypeT1Webusb     DeviceType = 1
	TypeT1WebusbBoot DeviceType = 2
	TypeT2           DeviceType = 3
	TypeT2Boot       DeviceType = 4
	TypeEmulator     DeviceType = 5
)

type USBInfo struct {
	Path      string
	VendorID  int
	ProductID int
	Type      DeviceType
	Debug     bool // has debug enabled?
}

type USBDevice interface {
	io.ReadWriter
	Close(disconnected bool) error
}

type session struct {
	path       string
	id         string
	dev        USBDevice
	call       int32 // atomic
	readMutex  sync.Mutex
	writeMutex sync.Mutex
}

type EnumerateEntry struct {
	Path    string     `json:"path"`
	Vendor  int        `json:"vendor"`
	Product int        `json:"product"`
	Type    DeviceType `json:"-"`     // used only in status page, not in JSON
	Debug   bool       `json:"debug"` // has debug enabled?

	Session      *string `json:"session"`
	DebugSession *string `json:"debugSession"`
}

type EnumerateEntries []EnumerateEntry

func (entries EnumerateEntries) Len() int {
	return len(entries)
}
func (entries EnumerateEntries) Less(i, j int) bool {
	return entries[i].Path < entries[j].Path
}
func (entries EnumerateEntries) Swap(i, j int) {
	entries[i], entries[j] = entries[j], entries[i]
}

type Core struct {
	bus USBBus

	normalSessions sync.Map
	debugSessions  sync.Map
	libusbMutex    sync.Mutex

	allowStealing bool
	reset         bool

	// We cannot make calls and enumeration at the same time,
	// because of some libusb/hidapi issues.
	// However, it is easier to fix it here than in the usb/ packages,
	// because they don't know about whole messages and just send
	// small packets by read/write
	//
	// Those variables help with that
	callsInProgress int
	callMutex       sync.Mutex
	lastInfosMutex  sync.RWMutex
	lastInfos       []USBInfo // when call is in progress, use saved info for enumerating
	// note - both lastInfos and Enumerate result have "paths"
	// as *fake paths* 2.0.26 onwards; just using device IDs,
	// unique for the device

	// the paths we present out are not actual paths
	// from 2.0.26 onwards;
	// it's just an ID of a device
	// we keep the IDs here
	usbPaths  map[int]string // id => path
	biggestID int

	log *memorywriter.MemoryWriter

	latestSessionID int
}

var (
	ErrWrongPrevSession = errors.New("wrong previous session")
	ErrSessionNotFound  = errors.New("session not found")
	ErrMalformedData    = errors.New("malformed data")
	ErrOtherCall        = errors.New("other call in progress")
)

const (
	VendorT1            = 0x534c
	ProductT1Firmware   = 0x0001
	VendorT2            = 0x1209
	ProductT2Bootloader = 0x53C0
	ProductT2Firmware   = 0x53C1
)

func New(bus USBBus, log *memorywriter.MemoryWriter, allowStealing, reset bool) *Core {
	c := &Core{
		bus:           bus,
		log:           log,
		allowStealing: allowStealing,
		reset:         reset,
		usbPaths:      make(map[int]string),
	}
	go c.backgroundListen()
	return c
}

const (
	iterMax   = 600
	iterDelay = 500 // ms
)

// This is here just to force recomputing the IDs of
// the disconnected devices (c.usbPaths)
// Note - this does not do anything when no device is connected
// or when no enumerate is run first...
// -> it runs whenever someone calls Enumerate/Listen
// and there are some devices left
// It does not spam USB that much more than listen itself
func (c *Core) backgroundListen() {
	for {
		time.Sleep(iterDelay * time.Millisecond)

		c.lastInfosMutex.RLock()
		linfos := len(c.lastInfos)
		c.lastInfosMutex.RUnlock()
		if linfos > 0 {
			c.log.Log("background enum runs")
			_, err := c.Enumerate()
			if err != nil {
				// we dont really care here
				c.log.Log("error - " + err.Error())
			}
		}
	}
}

func (c *Core) saveUsbPaths(devs []USBInfo) []USBInfo {
	for _, dev := range devs {
		add := true
		for _, usbPath := range c.usbPaths {
			if dev.Path == usbPath {
				add = false
			}
		}
		if add {
			newID := c.biggestID + 1
			c.biggestID = newID
			c.usbPaths[newID] = dev.Path
		}
	}
	reverse := make(map[string]string)
	for id, usbPath := range c.usbPaths {
		discard := true
		for _, dev := range devs {
			if dev.Path == usbPath {
				discard = false
			}
		}
		if discard {
			delete(c.usbPaths, id)
		} else {
			reverse[usbPath] = strconv.Itoa(id)
		}
	}
	res := make([]USBInfo, 0, len(devs))
	for _, dev := range devs {
		res = append(res, USBInfo{
			Path:      reverse[dev.Path],
			VendorID:  dev.VendorID,
			ProductID: dev.ProductID,
			Type:      dev.Type,
			Debug:     dev.Debug,
		})
	}
	return res
}

func (c *Core) Enumerate() ([]EnumerateEntry, error) {

	// avoid enumerating while acquiring the device
	// https://github.com/detahard/detahardd-go/issues/221
	c.libusbMutex.Lock()
	defer c.libusbMutex.Unlock()

	// Lock for atomic access to s.callInProgress.  It needs to be over
	// whole function, so that call does not actually start while
	// enumerating.
	c.callMutex.Lock()
	defer c.callMutex.Unlock()

	c.lastInfosMutex.Lock()
	defer c.lastInfosMutex.Unlock()

	// Use saved info if call is in progress, otherwise enumerate.
	infos := c.lastInfos

	c.log.Log(fmt.Sprintf("callsInProgress %d", c.callsInProgress))
	if c.callsInProgress == 0 {
		c.log.Log("bus")
		busInfos, err := c.bus.Enumerate()
		if err != nil {
			return nil, err
		}
		infos = c.saveUsbPaths(busInfos)
		c.lastInfos = infos
	}

	entries := c.createEnumerateEntries(infos)
	c.log.Log("release disconnected")
	c.releaseDisconnected(infos, false)
	c.releaseDisconnected(infos, true)
	return entries, nil
}

func (c *Core) findSession(e *EnumerateEntry, path string, debug bool) {
	s := (c.sessions(debug))
	s.Range(func(_, v interface{}) bool {
		ss := v.(*session)
		if ss.path == path {
			// Copying to prevent overwriting on Acquire and
			// wrong comparison in Listen.
			ssidCopy := ss.id
			if debug {
				e.DebugSession = &ssidCopy
			} else {
				e.Session = &ssidCopy
			}
			return false
		}
		return true
	})
}

func (c *Core) createEnumerateEntry(info USBInfo) EnumerateEntry {
	e := EnumerateEntry{
		Path:    info.Path,
		Vendor:  info.VendorID,
		Product: info.ProductID,
		Type:    info.Type,
		Debug:   info.Debug,
	}
	c.findSession(&e, info.Path, false)
	c.findSession(&e, info.Path, true)
	return e
}

func (c *Core) createEnumerateEntries(infos []USBInfo) EnumerateEntries {
	entries := make(EnumerateEntries, 0, len(infos))
	for _, info := range infos {
		e := c.createEnumerateEntry(info)
		entries = append(entries, e)
	}
	entries.Sort()
	return entries
}

func (entries EnumerateEntries) Sort() {
	sort.Sort(entries)
}

func (c *Core) sessions(debug bool) *sync.Map {
	if debug {
		return &c.debugSessions
	}
	return &c.normalSessions
}

func (c *Core) releaseDisconnected(infos []USBInfo, debug bool) {
	s := c.sessions(debug)
	s.Range(func(k, v interface{}) bool {
		ssid := k.(string)
		ss := v.(*session)
		connected := false
		for _, info := range infos {
			if ss.path == info.Path {
				connected = true
				break
			}
		}
		if !connected {
			c.log.Log(fmt.Sprintf("disconnected device %s", ssid))
			err := c.release(ssid, true, debug)
			// just log if there is an error
			// they are disconnected anyway
			if err != nil {
				c.log.Log(fmt.Sprintf("Error on releasing disconnected device: %s", err))
			}
		}
		return true
	})
}

func (c *Core) Release(session string, debug bool) error {
	return c.release(session, false, debug)
}

func (c *Core) release(
	ssid string,
	disconnected bool,
	debug bool,
) error {
	c.log.Log(fmt.Sprintf("session %s", ssid))
	s := c.sessions(debug)
	v, ok := s.Load(ssid)
	if !ok {
		c.log.Log("session not found")
		return ErrSessionNotFound
	}
	s.Delete(ssid)
	acquired := v.(*session)
	c.log.Log("bus close")
	err := acquired.dev.Close(disconnected)
	return err
}

func (c *Core) Listen(entries []EnumerateEntry, ctx context.Context) ([]EnumerateEntry, error) {
	c.log.Log("start")

	EnumerateEntries(entries).Sort()

	for i := 0; i < iterMax; i++ {
		c.log.Log("before enumerating")
		e, enumErr := c.Enumerate()
		if enumErr != nil {
			return nil, enumErr
		}
		for i := range e {
			e[i].Type = 0 // type is not exported/imported to json
		}
		if reflect.DeepEqual(entries, e) {
			c.log.Log("equal, waiting")
			select {
			case <-ctx.Done():
				c.log.Log(fmt.Sprintf("request closed (%s)", ctx.Err().Error()))
				return nil, nil
			default:
				time.Sleep(iterDelay * time.Millisecond)
			}
		} else {
			c.log.Log("different")
			entries = e
			break
		}
	}
	c.log.Log("encoding and exiting")
	return entries, nil
}

func (c *Core) findPrevSession(path string, debug bool) string {
	s := c.sessions(debug)
	res := ""
	s.Range(func(_, v interface{}) bool {
		ss := v.(*session)
		if ss.path == path {
			res = ss.id
			return false
		}
		return true
	})
	return res
}

func (c *Core) Acquire(
	path, prev string,
	debug bool,
) (string, error) {

	// avoid enumerating while acquiring the device
	// https://github.com/detahard/detahardd-go/issues/221
	c.libusbMutex.Lock()
	defer c.libusbMutex.Unlock()

	// note - path is *fake path*, basically device ID,
	// because that is what enumerate returns;
	// we convert it to actual path for USB layer

	c.log.Log(fmt.Sprintf("input path %s prev %s", path, prev))

	prevSession := c.findPrevSession(path, debug)

	c.log.Log(fmt.Sprintf("actually previous %s", prevSession))

	if prevSession != prev {
		return "", ErrWrongPrevSession
	}

	if (!c.allowStealing) && prevSession != "" {
		return "", ErrOtherCall
	}

	if prev != "" {
		c.log.Log("releasing previous")
		err := c.release(prev, false, debug)
		if err != nil {
			return "", err
		}
	}

	// reset device ONLY if no call on the other port
	// otherwise, USB reset stops other call
	otherSession := c.findPrevSession(path, !debug)
	reset := otherSession == "" && c.reset

	pathI, err := strconv.Atoi(path)
	if err != nil {
		return "", err
	}

	usbPath, exists := c.usbPaths[pathI]
	if !exists {
		return "", errors.New("device not found")
	}

	c.log.Log("trying to connect")
	dev, err := c.tryConnect(usbPath, debug, reset)
	if err != nil {
		return "", err
	}

	id := c.newSession(debug)

	sess := &session{
		path: path,
		dev:  dev,
		call: 0,
		id:   id,
	}

	c.log.Log(fmt.Sprintf("new session is %s", id))

	s := c.sessions(debug)
	s.Store(id, sess)

	return id, nil
}

// Chrome tries to read from detahard immediately after connecting,
// ans so do we.  Bad timing can produce error on s.bus.Connect.
// Try 3 times with a 100ms delay.
func (c *Core) tryConnect(path string, debug bool, reset bool) (USBDevice, error) {
	tries := 0
	for {
		c.log.Log(fmt.Sprintf("try number %d", tries))
		dev, err := c.bus.Connect(path, debug, reset)
		if err != nil {
			if tries < 3 {
				c.log.Log("sleeping")
				tries++
				time.Sleep(100 * time.Millisecond)
			} else {
				c.log.Log("tryConnect - too many times, exiting")
				return nil, err
			}
		} else {
			return dev, nil
		}
	}
}

func (c *Core) newSession(debug bool) string {
	c.latestSessionID++
	res := strconv.Itoa(c.latestSessionID)
	if debug {
		res = "debug" + res
	}
	return res
}

type CallMode int

const (
	CallModeRead      CallMode = 0
	CallModeWrite     CallMode = 1
	CallModeReadWrite CallMode = 2
)

func (c *Core) Call(
	body []byte,
	ssid string,
	mode CallMode,
	debug bool,
	ctx context.Context,
) ([]byte, error) {

	c.callMutex.Lock()
	c.callsInProgress++
	c.callMutex.Unlock()

	defer func() {
		c.callMutex.Lock()
		c.callsInProgress--
		c.callMutex.Unlock()
	}()

	s := c.sessions(debug)
	v, ok := s.Load(ssid)
	if !ok {
		return nil, ErrSessionNotFound
	}

	acquired := v.(*session)

	if mode != CallModeWrite {
		// This check is implemented only for /call and /read:
		// - /call: Two /calls should not run concurrently. Otherwise the "message writes" and "message reads"
		//   could interleave and the second caller would read the first response.
		// - /read: Although this could be possible we do not have a use case for that at the moment.
		// The check IS NOT implemented for /post, meaning /post can write even though some /call or /read
		// is in progress (but there are some read/write locks later on).

		c.log.Log("checking other call on same session")
		freeToCall := atomic.CompareAndSwapInt32(&acquired.call, 0, 1)
		if !freeToCall {
			return nil, ErrOtherCall
		}

		c.log.Log("checking other call on same session done")
		defer func() {
			atomic.StoreInt32(&acquired.call, 0)
		}()
	}

	finished := make(chan bool, 1)
	defer func() {
		finished <- true
	}()

	go func() {
		select {
		case <-finished:
			return
		case <-ctx.Done():
			c.log.Log(fmt.Sprintf("detected request close %s, auto-release", ctx.Err().Error()))
			errRelease := c.release(ssid, false, debug)
			if errRelease != nil {
				// just log, since request is already closed
				c.log.Log(fmt.Sprintf("Error while releasing: %s", errRelease.Error()))
			}
		}
	}()

	c.log.Log("before actual logic")
	bytes, err := c.readWriteDev(body, acquired, mode)
	c.log.Log("after actual logic")

	return bytes, err
}

func (c *Core) writeDev(body []byte, device io.Writer) error {
	c.log.Log("decodeRaw")
	msg, err := c.decodeRaw(body)
	if err != nil {
		return err
	}

	c.log.Log("writeTo")
	_, err = msg.WriteTo(device)
	return err
}

func (c *Core) readDev(device io.Reader) ([]byte, error) {
	c.log.Log("readFrom")
	msg, err := wire.ReadFrom(device, c.log)
	if err != nil {
		return nil, err
	}

	c.log.Log("encoding back")
	return c.encodeRaw(msg)
}

func (c *Core) readWriteDev(
	body []byte,
	acquired *session,
	mode CallMode,
) ([]byte, error) {

	if mode == CallModeRead {
		if len(body) != 0 {
			return nil, errors.New("non-empty body on read mode")
		}
		c.log.Log("skipping write")
	} else {
		acquired.writeMutex.Lock()
		err := c.writeDev(body, acquired.dev)
		acquired.writeMutex.Unlock()
		if err != nil {
			return nil, err
		}
	}

	if mode == CallModeWrite {
		c.log.Log("skipping read")
		return []byte{0}, nil
	}
	acquired.readMutex.Lock()
	defer acquired.readMutex.Unlock()
	return c.readDev(acquired.dev)
}

func (c *Core) decodeRaw(body []byte) (*wire.Message, error) {
	c.log.Log("readAll")

	c.log.Log("decodeString")

	if len(body) < 6 {
		c.log.Log("body too short")
		return nil, ErrMalformedData
	}

	kind := binary.BigEndian.Uint16(body[0:2])
	size := binary.BigEndian.Uint32(body[2:6])
	data := body[6:]
	if uint32(len(data)) != size {
		c.log.Log("wrong data length")
		return nil, ErrMalformedData
	}

	if wire.Validate(data) != nil {
		c.log.Log("invalid data")
		return nil, ErrMalformedData
	}

	c.log.Log("returning")
	return &wire.Message{
		Kind: kind,
		Data: data,

		Log: c.log,
	}, nil
}

func (c *Core) encodeRaw(msg *wire.Message) ([]byte, error) {
	c.log.Log("start")
	var header [6]byte
	data := msg.Data
	kind := msg.Kind
	size := uint32(len(msg.Data))

	binary.BigEndian.PutUint16(header[0:2], kind)
	binary.BigEndian.PutUint32(header[2:6], size)

	res := append(header[:], data...)

	return res, nil
}
